src/ore/src/safe_select.rs

```diff
@@ -0,0 +1,230 @@
+
+        // Fallback expression used when all select branches have been disabled.
+        ; $else:expr
+
+    }) => {{
+        tokio::select! {
+            $($biased;)?
+            $($bind = $crate::safe_select::assert_cancel_safe($fut), if $c => {$handle}),+
+            else => $else
+        }
+    }};
+
+    // ==== Normalize =====
+
+    // These rules match a single `safe_select!` branch and normalize it for processing by the first rule.
+
+    (@ { $biased:tt $($t:tt)* } ) => {
+        // No `else` branch
+        $crate::safe_select!(@{ $biased $($t)*; panic!("all branches are disabled and there is no else branch") })
+    };
+    (@ { $biased:tt $($t:tt)* } else => $else:expr $(,)?) => {
+        $crate::safe_select!(@{ $biased $($t)*; $else })
+    };
+    (@ { $biased:tt $($t:tt)* } $p:pat = $f:expr, if $c:expr => $h:block, $($r:tt)* ) => {
+        $crate::safe_select!(@{ $biased $($t)* $p = $f, if $c => $h, } $($r)*)
+    };
+    (@ { $biased:tt $($t:tt)* } $p:pat = $f:expr => $h:block, $($r:tt)* ) => {
+        $crate::safe_select!(@{ $biased $($t)* $p = $f, if true => $h, } $($r)*)
+    };
+    (@ { $biased:tt $($t:tt)* } $p:pat = $f:expr, if $c:expr => $h:block $($r:tt)* ) => {
+        $crate::safe_select!(@{ $biased $($t)* $p = $f, if $c => $h, } $($r)*)
+    };
+    (@ { $biased:tt $($t:tt)* } $p:pat = $f:expr => $h:block $($r:tt)* ) => {
+        $crate::safe_select!(@{ $biased $($t)* $p = $f, if true => $h, } $($r)*)
+    };
+    (@ { $biased:tt $($t:tt)* } $p:pat = $f:expr, if $c:expr => $h:expr ) => {
+        $crate::safe_select!(@{ $biased $($t)* $p = $f, if $c => $h, })
+    };
+    (@ { $biased:tt $($t:tt)* } $p:pat = $f:expr => $h:expr ) => {
+        $crate::safe_select!(@{ $biased $($t)* $p = $f, if true => $h, })
+    };
+    (@ { $biased:tt $($t:tt)* } $p:pat = $f:expr, if $c:expr => $h:expr, $($r:tt)* ) => {
+        $crate::safe_select!(@{ $biased $($t)* $p = $f, if $c => $h, } $($r)*)
+    };
+    (@ { $biased:tt $($t:tt)* } $p:pat = $f:expr => $h:expr, $($r:tt)* ) => {
+        $crate::safe_select!(@{ $biased $($t)* $p = $f, if true => $h, } $($r)*)
+    };
+
+    // ===== Entry points =====
+
+    (biased; $p:pat = $($t:tt)* ) => { $crate::safe_select!(@{ (biased) } $p = $($t)*) };
+    ( $p:pat = $($t:tt)* ) => { $crate::safe_select!(@{ () } $p = $($t)*) };
+    () => { compile_error!("safe_select! requires at least one branch.") };
+}
+
+use std::ops::{Deref, DerefMut};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// Marker trait that asserts a future is safe to cancel
+pub trait CancellationSafe: Future {}
+
+pub fn assert_cancel_safe<F: CancellationSafe>(f: F) -> F {
+    f
+}
+
+/// A type that wraps a future to mark it as cancellation safe. This is useful in situations where
+/// implementing the [CancellationSafe] trait is impossible (for example if it's an `impl Future`
+/// return type of a function
+pub struct CancelSafe<F>(F);
```

Comment:
this should be in a private module, and therefore un-nameable, otherwise it will just be abused

Reply:
alternatively, used the `Sealed` pattern to seal `CancellationSafe`

Reply:
Marked the trait unsafe and also added an unsafe `CancelSafe::new_unchecked` method